UpdateVcfSequenceDictionary handles stdout

ronlevine opened the pull request:
### Description

Implements the handling of stdout in https://github.com/broadinstitute/picard/issues/507.
If `stdout` is the output file, set to write to a stream instead of a file.

----

### Checklist (never delete this)

Never delete this, it is our record that procedure was followed. If you find that for whatever reason one of the checklist points doesn't apply to your PR, you can leave it unchecked but please add an explanation below.

#### Content
- [X] Added or modified tests to cover changes and any new functionality
- [ ] Edited the README / documentation (if applicable)
- [X] All tests passing on Travis

#### Review
- [ ] Final thumbs-up from reviewer
- [X] Rebase, squash and reword as applicable

For more detailed guidelines, see https://github.com/broadinstitute/picard/wiki/Guidelines-for-pull-requests

ronlevine commented:
@yfarjoun Please review.

nh13 commented on src/main/java/picard/vcf/UpdateVcfSequenceDictionary.java:
```diff
@@ -59,7 +59,7 @@
     public File INPUT;
 
-    @Option(shortName = StandardOptionDefinitions.OUTPUT_SHORT_NAME, doc = "Output VCF to be written.")
+    @Option(shortName = StandardOptionDefinitions.OUTPUT_SHORT_NAME, doc = "Output VCF to be written to a file or stdout")
```
Not needed.

ronlevine replied:
Reverted the change.

nh13 commented on src/main/java/picard/vcf/UpdateVcfSequenceDictionary.java:
```diff
@@ -89,7 +91,15 @@ protected int doWork() {
 
-        final VariantContextWriter vcfWriter = builder.setOutputFile(OUTPUT).build();
+        // handle stdout
+        if ( OUTPUT.getPath().equals(STD_OUT) ) {
```
Stdout would be specified with `O=/dev/stdout`.  How about:
`builder.setOutputStream(new FileOutputStream(OUTPUT, false));`
instead of the if/else?

ronlevine replied:
That's cleaner but how can the stdout be redirected to a file for validation by a test? With `System.out`, it's easy by using `System.setOut(PrintStream out)`.

ronlevine replied:
@yfarjoun I like the code change suggestion but it would it would be tough to validate the output. I can bypass that check in my test.Thoughts?

yfarjoun replied:
I don't think you need the if statement here at all, as @nh13 said. please remove it and wrap OUTPUT with a FileOutputStream (and remove the STD_OUT constant)

ronlevine replied:
@yfarjoun The output cannot be redirected to a file via `System.setOut()`. the stream must be `System.out`, `builder.setOutputStream(System.out)`.

ronlevine replied:
Done.

nh13 commented on src/test/java/picard/vcf/UpdateVcfSequenceDictionaryTest.java:
```diff
@@ -28,38 +28,61 @@
 public class UpdateVcfSequenceDictionaryTest {
     private static final File TEST_DATA_PATH = new File("testdata/picard/vcf/");
     private static final File OUTPUT_DATA_PATH = IOUtil.createTempDir("UpdateVcfSequenceDictionaryTest", null);
+    private static final File STD_OUT_FILE = new File(OUTPUT_DATA_PATH, "stdout.vcf");
```
I am not so sure about these tests, @yfarjoun

ronlevine replied:
The intent is to redirect stdout into a file, so it could be verified. Do you have a better idea for verification?

nh13 replied:
Sorry, I meant to say I didn't review the tests, sorry for the confusion.

yfarjoun replied:
I think you should be calling `System.setOut(PrintStream out)` from the test and not modify the code in UpdateVcfSequenceDictionary. 

so the test would be something like 
```java
File STD_OUT_FILE = new File(OUTPUT_DATA_PATH, "stdout.vcf");
Stream stdout = new OutputStream(STD_OUT_FILE);
System.setOut(stdout);

//run test

//close stream if needed

//examine file.
```

ronlevine commented:
@nh13 @yfarjoun I cleaned up the test and removed the "if" statement so that the OUTPUT is a stream. Removing that logic makes the code more concise and flexible. Note that `setOutputFile` is just and adapter, it converts the file to a stream. Let  know if you have any more comments.

yfarjoun commented:
something isn't right. 

The value of OUTPUT has to be the string "/dev/stdout"

Then you should be setting the output stream to point to a file using System.setOut()

ronlevine commented:
@yfarjoun How about this? OUTPUT=/dev/stdout.

yfarjoun commented on src/test/java/picard/vcf/UpdateVcfSequenceDictionaryTest.java:
```diff
@@ -28,37 +28,58 @@
         outputFile.deleteOnExit();
 
+        // Stream standard output a file
+        if ( outputFileName.equals(UpdateVcfSequenceDictionary.STD_OUT) ) {
```
use a local variable for STD_OUT.

yfarjoun replied:
same here (fewer spaces)

yfarjoun commented on src/test/java/picard/vcf/UpdateVcfSequenceDictionaryTest.java:
```diff
@@ -28,37 +28,58 @@
 
         Assert.assertEquals(updateVcfSequenceDictionary.instanceMain(new String[0]), 0);
 
+        if ( outputFileName.equals(UpdateVcfSequenceDictionary.STD_OUT) ) {
```
a bit fewer spaces
```
if (outputFileName.equals(UpdateVcfSequenceDictionary.STD_OUT)) {
```

yfarjoun commented:
back to you @ronlevine

yfarjoun commented:
@ronlevine I have uploaded a branch with hat I would imagine would be a "clean" solution to the following branch (sorry about the `)` in the branch name...typo) : https://github.com/broadinstitute/picard/tree/yf)uvsd_streams

ronlevine commented:
@yfarjoun Please take a look at my changes. I took a look at your branch and incorporated the file stream redirection by file descriptor for stdout. Thanks, that solved the file validation problem. Note that the only the `setOutputStream` is needed and the if statement can be removed. I also found that if the test was run with `STD_OUT_NAME` as the first or only  `outputFileName` , it would fail while writing the header:
```
htsjdk.samtools.util.RuntimeIOException: Couldn't write file java.io.FileOutputStream@7d907bac

	at htsjdk.variant.variantcontext.writer.VCFWriter.writeHeader(VCFWriter.java:139)
```
because
`final FileOutputStream stream = new FileOutputStream(STD_OUT_FILE)` goes out of scope and is destroyed. So, I moved it's declaration outside the if statement.

yfarjoun commented on src/test/java/picard/vcf/UpdateVcfSequenceDictionaryTest.java:
```diff
@@ -28,37 +28,69 @@
 
         final UpdateVcfSequenceDictionary updateVcfSequenceDictionary = new UpdateVcfSequenceDictionary();
+
+        FileOutputStream stream = null;  // placed outside of if statement so it's not destroyed before executing updateVcfSequenceDictionary
```
can be final. (just set to null in the else branch)

ronlevine replied:
Done.

yfarjoun commented on src/test/java/picard/vcf/UpdateVcfSequenceDictionaryTest.java:
```diff
@@ -28,37 +28,69 @@
+
+            stream = new FileOutputStream(STD_OUT_FILE);
+
+            // Ugliness required to write to m a stream given as a string on the commandline.
```
is that `m` in my original comment? in any case, it's a typo, please fix.

ronlevine replied:
Nope, just a case of my fat fingers when changing `read from` to `write to`.

ronlevine commented:
Squashed and rebased.

ronlevine commented:
@yfarjoun OK to merge?

yfarjoun commented:
yes. I'm fine with it. 👍  from me. 

In the future, please refrain from squashing your commits until the merge (you can squash during the final merge itself), so that your reviewers can have an easier time viewing the latest changes you made.
